src/manual/devel/best-practices/microservices.md

```diff
@@ -0,0 +1,307 @@
+
+```js
+let server = await load('server', {
+  profile: 'test',
+  dependency1: fakeDep1,
+  dependency2: fakeDep2,
+});
+```
+
+For tests that must have credentials, check for the presence of credentials and mark the suite as pending if they are not available:
+
+```js
+suite("things", function() {
+  if (!helper.haveRealCredentials) {
+    this.pending = true;
+  }
+});
+```
+
+This will generate clear output for anyone running the tests without credentials, showing that many tests were not run.
+If they make a pull request, then the full suite will run in automation, and any issues not detected locally will be revealed.
+
+## Taskcluster Libraries
+
+### General
+
+Do not use `taskcluster-lib-base`.
+Instead, depend directly on the `taskcluster-lib-*` libraries the service requires.
+
+The following sections describe best practices for specific platform libraries.
+
+### taskcluster-lib-loader
+
+The main entry-point for the service should be `src/main.js`, which should use [taskcluster-lib-loader](https://github.com/taskcluster/taskcluster-lib-loader) and have the following initialization code:
+
+```js
+// If this file is executed launch component from first argument
+if (!module.parent) {
+  load(process.argv[2], {
+    process: process.argv[2],
+    profile: process.env.NODE_ENV,
+  }).catch(err => {
+    console.log(err.stack);
+    process.exit(1);
+  });
+}
+
+// Export load for tests
+module.exports = load;
+```
+
+Entries in `Procfile`, then, look like `web: node lib/main.js server`.
+The web service should always be the component named `server`.
+All services, including those run from the Heroku scheduler, should start like this, via `lib/main.js`.
+
+### azure-entities
+
+Each Azure table should be defined in `src/data.js` using a cascade of `configure` calls, one for each version:
+
+```js
+var MyEntity = Entity.configure({
+  version:              1,
+}).configure({
+  version:              2,
+  migrate: function(item) {
+    // ...
+  },
+}).configure({
+  version:              3,
+  migrate: function(item) {
+    // ...
+  },
+});
+```
+
+The result is a `MyEntity` class that can be setup with additional configuration in a loader component, in `src/main.js`:
+
+```js
+{
+  MyEntity: {
+    requires: ['cfg', 'process', 'monitor'],
+    setup: ({cfg, process, monitor}) => {
+      return data.MyEntity.setup(_.defaults({
+        table:        cfg.app.hookTable,
+        monitor:      monitor.prefix(cfg.app.hookTable.toLowerCase()),
+        component:    cfg.app.component,
+        process,
+      }, cfg.azureTable, cfg.taskcluster));
+    },
+  },
+}
+```
+
+In `src/data.js`, it is common to add utility methods for the entity type.
+For entities which can be fetched via an API, a `json` method is common:
+
+```js
+MyEntity.prototype.json = () => {
+  return {
+    foo: this.foo,
+  };
+};
+```
+
+### taskcluster-lib-api
+
+The API definition should be in `src/v1.js`:
```

Comment:
Yeah, thinks seem split fairly evenly on this point.  I'll say "..or `src/api.js`".  And again, we can make a PR-and-cleanup later :)